### PR TITLE
Add Cognito client provider

### PR DESCRIPTION
**Pull request by emanuelecastelli**

### CODE REVIEW

**Comment by emanuelecastelli** [comment]
Hi @kevinchalet ,
is it possible to know the release times of this PR? We have to publish in production within two weeks and I would like to understand if we can already use the main repository or our fork.
Thank you

File: src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml [hunk 1]
```diff
@@ -189,4 +189,12 @@
     </Setting>
   </Provider>
 
+ <Provider Name="Cognito" Documentation="https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-reference.html">
```
Comment (kevinchalet) [inline src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml]: Can you please move it so the alphabetical order is respected? (see https://documentation.openiddict.com/guides/contributing-a-new-web-provider.html#add-a-new-provider-node-for-the-new-provider)

File: src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml [hunk 2]
```diff
@@ -189,4 +189,12 @@
...
+    <Environment Issuer="https://cognito-idp.{region}.amazonaws.com/{userPoolId}"/>
+    <Setting PropertyName="Region" ParameterName="region" Type="String" Required="true"
+           Description="The AWS region" />
+    <Setting PropertyName="UserPoolId" ParameterName="userPoolId" Type="String" Required="true"
+           Description="The AWS region" />
```
Comment (kevinchalet) [inline src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml]: I suspect this description is incorrect.

File: src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml [hunk 3]
```diff
@@ -189,4 +189,12 @@
     </Setting>
   </Provider>
 
+ <Provider Name="Cognito" Documentation="https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-reference.html">
+    <Environment Issuer="https://cognito-idp.{region}.amazonaws.com/{userPoolId}"/>
+    <Setting PropertyName="Region" ParameterName="region" Type="String" Required="true"
+           Description="The AWS region" />
+    <Setting PropertyName="UserPoolId" ParameterName="userPoolId" Type="String" Required="true"
```
Comment (kevinchalet) [inline src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml]: Consider changing the `ParameterName` to just `identifier` (it's only used in the *Builder methods, so don't change the `{userPoolId}` in the issuer URL format).

File: src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml [hunk 4]
```diff
@@ -189,4 +189,12 @@
     </Setting>
   </Provider>
 
+ <Provider Name="Cognito" Documentation="https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-reference.html">
+    <Environment Issuer="https://cognito-idp.{region}.amazonaws.com/{userPoolId}"/>
```
Comment (kevinchalet) [inline src/OpenIddict.Client.WebIntegration/OpenIddictClientWebIntegrationProviders.xml]: Please add a space before the self-closing tag.

**Review by kevinchalet — COMMENTED** [review]
Thanks for your contribution 😃

**Comment by kevinchalet** [comment]
Hi @emanuelecastelli,

I plan to release a first RC later this month and the RTM version should ship next month.

Cheers.

File: sandbox/OpenIddict.Sandbox.AspNetCore.Client/Startup.cs [hunk 5]
```diff
@@ -148,6 +149,14 @@ public void ConfigureServices(IServiceCollection services)
                            options.SetClientId("bXgwc0U3N3A3YWNuaWVsdlRmRWE6MTpjaQ")
                                   .SetClientSecret("VcohOgBp-6yQCurngo4GAyKeZh0D6SUCCSjJgEo1uRzJarjIUS")
                                   .SetRedirectUri("https://localhost:44381/callback/login/twitter");
+                       })
+                       .UseCognito(options =>
+                       {
+                           options.SetClientId("COGNITO_CLIENT_ID")
```
Comment (kevinchalet) [inline sandbox/OpenIddict.Sandbox.AspNetCore.Client/Startup.cs]: Note: if you don't plan to add "real" credentials, please remove the changes made to the sample (only providers with credentials that can be used as-is are expected to be added in the sandbox)

File: sandbox/OpenIddict.Sandbox.AspNetCore.Client/Controllers/AuthenticationController.cs [hunk 6]
```diff
@@ -20,7 +20,8 @@ public ActionResult LogIn(string provider, string returnUrl)
             !string.Equals(provider, Providers.GitHub,  StringComparison.Ordinal) &&
             !string.Equals(provider, Providers.Google,  StringComparison.Ordinal) &&
             !string.Equals(provider, Providers.Reddit,  StringComparison.Ordinal) &&
-            !string.Equals(provider, Providers.Twitter, StringComparison.Ordinal))
+            !string.Equals(provider, Providers.Twitter, StringComparison.Ordinal) &&
```
Comment (kevinchalet) [inline sandbox/OpenIddict.Sandbox.AspNetCore.Client/Controllers/AuthenticationController.cs]: Please remove any change to the sample as the Cognito provider is not going to be enabled by default.

**Comment by emanuelecastelli** [comment]
Hi @kevinchalet , any news about that? It's ok for merge?

**Comment by kevinchalet** [comment]
Hi. Nope, you forgot to remove the changes from `Startup.cs` 😄

**Comment by emanuelecastelli** [comment]
> Hi. Nope, you forgot to remove the changes from `Startup.cs` 😄

Oh my...  😅

**Comment by kevinchalet** [comment]
Looks like it's ready to merge (there's a git conflict, but I'll work around it, so don't worry).

Do you confirm the provider works fine? Any chance you could share a few screenshots so I have an idea of what AWS Cognito looks like? :smile:

**Comment by emanuelecastelli** [comment]
> Looks like it's ready to merge (there's a git conflict, but I'll work around it, so don't worry).
> 
> Do you confirm the provider works fine? Any chance you could share a few screenshots so I have an idea of what AWS Cognito looks like? 😄

Sure @kevinchalet , 
tomorrow I will probably be able to prepare some screenshots of the Cognito dashboard with the relevant settings. 
Consider that AWS provides you with the possibility to register an authentication client for free (in case you want to try it yourself).

**Comment by emanuelecastelli** [comment]
Hi @kevinchalet , 
here some screenshot taken from Cognito dashboard.
Do you need something else?
appclient
generalsettings
users

**Comment by kevinchalet** [comment]
Great, thanks @emanuelecastelli!

I found a valid user pool ID online and was able to retrieve the discovery document:

```json
{
  "authorization_endpoint": "https://cognito-idp.us-east-1.amazonaws.com/us-east-1_[redacted]/authorize",
  "id_token_signing_alg_values_supported": [
    "RS256"
  ],
  "issuer": "https://cognito-idp.us-east-1.amazonaws.com/us-east-1_[redacted]",
  "jwks_uri": "https://cognito-idp.us-east-1.amazonaws.com/us-east-1_[redacted]/.well-known/jwks.json",
  "response_types_supported": [
    "code",
    "token"
  ],
  "scopes_supported": [
    "openid",
    "email",
    "phone",
    "profile"
  ],
  "subject_types_supported": [
    "public"
  ],
  "token_endpoint": "https://cognito-idp.us-east-1.amazonaws.com/us-east-1_[redacted]/token",
  "token_endpoint_auth_methods_supported": [
    "client_secret_basic",
    "client_secret_post"
  ],
  "userinfo_endpoint": "https://cognito-idp.us-east-1.amazonaws.com/us-east-1_[redacted]/userInfo"
}
```

One important thing to note is that Cognito doesn't return a list of supported grant types. Once https://github.com/openiddict/openiddict-core/pull/1601 is merged, the Cognito integration won't work for the `client_credentials` and `refresh_token` grants.

I'll merge this PR as-is and add the necessary changes to amend the Cognito configuration. Once this change is added, would it be possible for you to test the latest `OpenIddict.Sandbox.AspNetCore.Client` bits to ensure the `refresh_token` flow works?

**Comment by kevinchalet** [comment]
@emanuelecastelli could you please clone https://github.com/kevinchalet/openiddict-core/tree/supported_grant_types_check and give it a try? By looking at the Cognito docs, a refresh_token should be returned for the code flow, without requiring a specific scope, so you should see the new `Refresh the access token` button in the sandbox demo after a successful Cognito authorization dance.

**Comment by emanuelecastelli** [comment]
@kevinchalet,
i can confirm that in `supported_grant_types_check` branch Cognito refresh token get returned correctly.
Thank you very much for your work and your support :)
Schermata 2022-11-28 alle 11 56 27

**Comment by kevinchalet** [comment]
Awesome, thanks @emanuelecastelli 😃 

> Thank you very much for your work and your support :)

Thanks for the kind words! If you use OpenIddict for professional, please don't hesitate to encourage your employer to sponsor the project to it can stay free in the future 😃

**Comment by emanuelecastelli** [comment]
> Awesome, thanks @emanuelecastelli 😃
> 
> > Thank you very much for your work and your support :)
> 
> Thanks for the kind words! If you use OpenIddict for professional, please don't hesitate to encourage your employer to sponsor the project to it can stay free in the future 😃

Yeah, for sure ;)

**Comment by emanuelecastelli** [comment]
Hi @kevinchalet ,
do you have any idea when you will release the next preview with Cognito support?
Thx

**Pull request merged**